docs/flow: Document component transforming sd targets

tpaschalis opened the pull request:
Signed-off-by: Paschalis Tsilias <paschalis.tsilias@grafana.com>

#### PR Description
This PR adds documentation for the recently added `discovery.transformer` component. It also renames it to **targets.mutate**.

~If we have reservations about the name, this might be a good place to kickstart that discussion.~

#### Which issue(s) this PR fixes
Fixes #1683

#### Notes to the Reviewer

Do you think we're okay to close #1683 after this?

#### PR Checklist

- [ ] CHANGELOG updated (N/A for now, until we tell people what Flow is about)
- [X] Documentation added
- [ ] Tests updated (N/A)

tpaschalis commented:
I'm not sure how we want to document custom object/structs. 

I'll take a look at Hashicorp's documentation to see how they handle Terraform modules and act accordingly.

marctc commented:
> 
> This PR adds documentation for the recently added `discovery.transformer` component.
> 
> If we have reservations about the name, this might be a good place to kickstart that discussion.

I'm not fully convinced about the name either. It feels that the `discovery` module would enable
developing other components to discover other resources than services. The name `sd` is not
very comprehensive either unless you now the context of Prometheus very well. I'd go for
`service-discovery`, it's long but it's descriptive.

> I'm not sure how we want to document custom object/structs.
> 
> I'll take a look at Hashicorp's documentation to see how they handle Terraform modules and act accordingly.

I don't recall seeing in any terraform's documentation page any reference of Go. I wouldn't add
those to documentation regardless. It feels more like an implementation detail that Flow users shouldn't
care.

rfratto commented on docs/flow/components/discovery.transformer.md:
```diff
@@ -0,0 +1,104 @@
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`targets` | `[]map[string]string` | The input targets | | no
```
Hmm, shouldn't targets be required though? Why have this component if you're not providing targets?

rfratto replied:
Oh, sorry, required means that the field has to be set by the user, even if it's the zero value. i.e., for required targets, this is still valid:

```hcl 
targets = []
```

but omitting the targets field isn't.

rfratto commented on docs/flow/components/discovery.transformer.md:
```diff
@@ -0,0 +1,104 @@
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`targets` | `[]map[string]string` | The input targets | | no
```
Let's use the type definitions Terraform uses here: `list(map(string))`. We'll need to revisit later to see if this is the type syntax we want to use long term, but something other than the Go type makes sense to me.

rfratto commented on docs/flow/components/discovery.transformer.md:
```diff
@@ -0,0 +1,104 @@
+that can be applied to an input target. If you want an overview of how
+relabeling works you can refer to the [Prometheus docs][] or this [blogpost][].
+
+#### Argument Reference
```
Drop the header here so there's only one "argument" section in the docs.

rfratto commented on docs/flow/components/discovery.transformer.md:
```diff
@@ -0,0 +1,104 @@
+
+Name | Type | Description
+---- | ---- | -----------
+`output_targets` | `[]map[string]string` | The resulting targets' label set.
```
nit: like the arguments, I think this should also be `list(map(string))`.

rfratto replied:
nit: For the description, I'd like it to be a little more descriptive: 

> The set of targets after applying transformations.

rfratto commented on docs/flow/components/discovery.transformer.md:
```diff
@@ -0,0 +1,104 @@
+---- | ---- | -----------
+`output_targets` | `[]map[string]string` | The resulting targets' label set.
+
+The number of targets exposed in `output_targets` will be less or equal than
+the number of input targets, as some of them may be dropped by the
+relabeling process.
```
It could be more too, if you use label mapping. I think we can just drop this paragraph.

tpaschalis replied:
I'm okay with dropping the paragraph, but can they actually be more? I thought that labelmap will just rename labels from a specific target.

rfratto replied:
It technically creates more. Given this label set: 

```
{ 
  "__label_foo" = "1", 
  "__label_bar" = "2",
}
```

And a relabel config of

```hcl
action      = "labelmap"
regex       = "__label_(.*)" 
replacement = "${1}" 
```

the resulting label set will be:

```
{ 
  "__label_foo" = "1", 
  "__label_bar" = "2", 
  "foo" = "1",
  "bar" = "2",
}
```

rfratto replied:
Ahh, you're right, my bad :)

rfratto commented on docs/flow/components/discovery.transformer.md:
```diff
@@ -0,0 +1,104 @@
+
+## Component health
+
+Any `discovery.transformer` component will be reported as healthy whenever
+it is able to correctly apply all relabeling steps to the input targets.
+If the either the input `targets`, or `relabel_config` blocks could not
+be parsed or the relabeling steps could not be applied, the component will
+be reported as unhealthy. In those cases, exported fields will be kept at
+the last healthy values.
```
I think we can shorten this quite a bit:

> The `discovery.transformer` component will only be reported as unhealthy when given an invalid configuration.

rfratto commented on docs/flow/components/discovery.transformer.md:
```diff
@@ -0,0 +1,104 @@
+    action        = "replace"
+  } 
+  
+  relabel_config {
+    source_labels = ["app"]
+    action = "keep"
+    regex  = "backend"
+  }
+}
```
This should to be reformatted to conform to the HCL formatting rules:

```suggestion
  relabel_config {
    source_labels = ["app"]
    action        = "keep"
    regex         = "backend"
  }
}
```

rfratto commented:
Something that just occurred to me: I think the syntax used for replacing regex capture groups doesn't work in HCL: `"${1}"` would be interpreted as an HCL interpolated expression and the user would need to escape with `$${1}`. Can we document that?

tpaschalis commented:
I've added a small note that users should refer to capture groups using either the `$1` or `$${1}` notations, and renamed the package to `targets.mutate`.

Let me know how it looks!

rfratto commented on docs/flow/components/targets.mutate.md:
```diff
@@ -0,0 +1,102 @@
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+targets | list(map(string)) | The targets to mutate. | | no
```
```suggestion
targets | list(map(string)) | The targets to mutate. | | **yes**
```

tpaschalis replied:
Whoops, thanks for the catch.

rfratto commented on docs/flow/components/targets.mutate.md:
```diff
@@ -0,0 +1,102 @@
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+targets | list(map(string)) | The targets to mutate. | | no
+relabel_config | RelabelConfig | The relabeling steps to apply. | | no
```
Hmm, for now, can we drop the block as being mentioned as an argument? I think blocks will have different documentation semantics than attributes.

```suggestion
```

tpaschalis replied:
Yeah, I wasn't sure how to handle them. Makes sense.

rfratto commented on docs/flow/components/targets.mutate.md:
```diff
@@ -0,0 +1,102 @@
+relabel_config | RelabelConfig | The relabeling steps to apply. | | no
+
+
+### RelabelConfig
```
And then change this to be about a block:

```suggestion
### `relabel_config` block
```

rfratto reviewed (commented):
Few nits about how we want to document inner blocks, but LGTM, thanks!
